homeassistant/components/aquacell/coordinator.py

```diff
@@ -0,0 +1,87 @@
+"""Coordinator to update data from Aquacell API."""
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from aioaquacell import (
+    AquacellApi,
+    AquacellApiException,
+    AuthenticationFailed,
+    Softener,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONF_REFRESH_TOKEN, UPDATE_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AquacellCoordinator(DataUpdateCoordinator[dict[str, Softener]]):
+    """My aquacell coordinator."""
+
+    config_entry: ConfigEntry
+
+    def __init__(self, hass: HomeAssistant, aquacell_api: AquacellApi) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Aquacell Coordinator",
+            update_interval=timedelta(seconds=UPDATE_INTERVAL),
+        )
+
+        self.refresh_token = self.config_entry.data[CONF_REFRESH_TOKEN]
+        self.email = self.config_entry.data[CONF_EMAIL]
+        self.password = self.config_entry.data[CONF_PASSWORD]
+        self.aquacell_api = aquacell_api
+
+    async def _async_update_data(self) -> dict[str, Softener]:
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables
+        so entities can quickly look up their data.
+        """
+
+        if self.refresh_token is None:
+            raise ConfigEntryError("No token available.")
+
+        async with asyncio.timeout(10):
+            _LOGGER.debug("Logging in using: %s", self.refresh_token)
+
+            try:
+                await self.aquacell_api.authenticate_refresh(self.refresh_token)
```

Comment:
What's the validity of a refresh token? Can we store that as well somewhere?

Reply:
It's currently set to 30 days, depends on their AWS configuration. Haven't seen this value in any of the authentication responses so decided to just attempt to reauthenticate whenever it fails to authenticate using the refresh token.

We could hardcode the 30 days and check on that, but it wouldn't be robuust against configuration changes on their cloud.

Reply:
What kind of token is it? a JWT?

Reply:
Yes, but the payload is encrypted.

Reply:
Do you have an example?

Reply:
`eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiUlNBLU9BRVAifQ.QZu553nflsb4EGa0mje9V8ySxd62MSwF6ceNpZnCvi_oTs3kIOdyoA7tbSbXaiNmlm1Rrfj-lMqJ36Cc7KKRrIfAKVvOnw6uccBYjwS-E9pn0hWQhlanLA8zPwlxNi9_lt2yhcNCvFvPEwFzmzj_-n7bOXvq0LNcNurv7tpdfA6ftfqvGj2vQHMxzIsDZ0KzrOsZlxEXoVfxuvk12RhwM1AG-oBJZ3FXpjGAnim-YIVNKWWRX35tUV4-Qhl5oMXPjb5ccSss-ONE7Ohkd-leC9eDhIw7osU_Ny6aP1kSOOO1qwq0exXOJwWWFqL_cHmL22Y80AcXdpHZAjJW7go_UA`

Reply:
Oh lol, i was like oh it's just a jwt, looks usable. But then indeed. Why would you encrypt this haha.

Hmm, i think we can maybe just roll with storing the date we requested the token and check if it exceeds 30 days. This makes the check if the token is valid a lot cheaper. It would only become a real problem when they make the validity 1 day for example. It would trigger reauth every day, but given that people watch issues and we release a hotfix about every week, we can then change this.

Wdyt?

Reply:
Oh wait, I forgot that we already have username and password, so we don't need to do reauth, unless the token retrieval changes. So I think this can be a very user friendly approach

Reply:
Yeah agreed, I'll add the 30 day check.